backport http://gcc.gnu.org/viewcvs?view=revision&revision=162404
from trunk

Which fixes http://gcc.gnu.org/bugzilla/show_bug.cgi?id=43698

2010-07-22  Ramana Radhakrishnan  <ramana.radhakrishnan@arm.com>

        PR target/43698
        * config/arm/arm.md: Split arm_rev into *arm_rev
        and *thumb1_rev. Set *arm_rev to be predicable.

2010-07-22  Ramana Radhakrishnan  <ramana.radhakrishnan@arm.com>

        PR target/43698
        * gcc.target/arm/pr43698.c: New test.


/scratch/oe/sysroots/i686-linux/usr/share/quilt/compat/date
Usage: date [OPTION]... [+FORMAT]
Display the current time in the given FORMAT.

  -d, --date=STRING         display time described by STRING, not `now'
  -f, --file=DATEFILE       like --date once for each line of DATEFILE
  -R, --rfc-822             output RFC-822 compliant date string
  -u, --utc, --universal    print or set Coordinated Universal Time
      --help                display this help and exit
date is /scratch/oe/sysroots/i686-linux/usr/share/quilt/compat/date
date is /scratch/oe/sysroots/i686-linux/usr/share/quilt/compat/date
date is /bin/date
date is /scratch/oe/sysroots/i686-linux/usr/share/quilt/compat/date
date is /scratch/oe/sysroots/i686-linux/usr/share/quilt/compat/date
date is /bin/date
date is /scratch/oe/sysroots/i686-linux/usr/share/quilt/compat/date
date is /scratch/oe/sysroots/i686-linux/usr/share/quilt/compat/date
date is /bin/date
date is /scratch/oe/sysroots/i686-linux/usr/share/quilt/compat/date
date is /scratch/oe/sysroots/i686-linux/usr/share/quilt/compat/date
date is /bin/date
date
Khem
Index: gcc-4.5/gcc/config/arm/arm.md
===================================================================
--- gcc-4.5.orig/gcc/config/arm/arm.md	2010-07-20 20:31:25.000000000 -0700
+++ gcc-4.5/gcc/config/arm/arm.md	2010-07-22 14:55:54.303169081 -0700
@@ -11197,15 +11197,21 @@
    (set_attr "length" "4")]
 )
 
-(define_insn "arm_rev"
+(define_insn "*arm_rev"
   [(set (match_operand:SI 0 "s_register_operand" "=r")
 	(bswap:SI (match_operand:SI 1 "s_register_operand" "r")))]
-  "TARGET_EITHER && arm_arch6"
-  "rev\t%0, %1"
-  [(set (attr "length")
-        (if_then_else (eq_attr "is_thumb" "yes")
-		      (const_int 2)
-		      (const_int 4)))]
+  "TARGET_32BIT && arm_arch6"
+  "rev%?\t%0, %1"
+  [(set_attr "predicable" "yes")
+   (set_attr "length" "4")]
+)
+
+(define_insn "*thumb1_rev"
+  [(set (match_operand:SI 0 "s_register_operand" "=l")
+	(bswap:SI (match_operand:SI 1 "s_register_operand" "l")))]
+  "TARGET_THUMB1 && arm_arch6"
+   "rev\t%0, %1"
+  [(set_attr "length" "2")]
 )
 
 (define_expand "arm_legacy_rev"
Index: gcc-4.5/gcc/testsuite/gcc.target/arm/pr43698.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ gcc-4.5/gcc/testsuite/gcc.target/arm/pr43698.c	2010-07-22 14:56:35.406670213 -0700
@@ -0,0 +1,39 @@
+/* { dg-do run } */
+/* { dg-options "-Os -march=armv7-a" } */
+#include <stdint.h>
+#include <stdlib.h>
+
+
+char do_reverse_endian = 0;
+
+#  define bswap_32(x) \
+  ((((x) & 0xff000000) >> 24) | \
+   (((x) & 0x00ff0000) >>  8) | \
+   (((x) & 0x0000ff00) <<  8) | \
+   (((x) & 0x000000ff) << 24))
+
+#define EGET(X) \
+  (__extension__ ({ \
+      uint64_t __res; \
+      if (!do_reverse_endian) {    __res = (X); \
+      } else if (sizeof(X) == 4) { __res = bswap_32((X)); \
+      } \
+      __res; \
+    }))
+
+void __attribute__((noinline)) X(char **phdr, char **data, int *phoff)
+{
+  *phdr = *data + EGET(*phoff);
+}
+
+int main()
+{
+  char *phdr;
+  char *data = (char *)0x40164000;
+  int phoff = 0x34;
+  X(&phdr, &data, &phoff);
+  if (phdr != (char *)0x40164034)
+    abort ();
+  exit (0);
+}
+
